### PR TITLE
fix: clean up async calls with expired leases

### LINKED ISSUE #2214: Async calls get stuck if controller dies
Steps to reproduce
- Comment out db cleanup code used in dev: https://github.com/TBD54566975/ftl/blob/main/backend/controller/sql/databasetesting/devel.go#L77
- Start FTL and trigger an async call that takes some time. Here's what I did:
    - Make [this verb](https://github.com/TBD54566975/ftl/blob/matt2e%2Fcatch/backend/controller/pubsub/testdata/go/subscriber/subscriber.go#L33) basically just do `time.Sleep(time.Minute); return nil`
    - `ftl dev --log-level=DEBUG backend/controller/pubsub/testdata/go/subscriber backend/controller/pubsub/testdata/go/publisher --recreate`
    - Go to the console and trigger the verb `publisher.publishOneToTopic2`
- Wait a few seconds so we know the async call has started
- Kill FTL (ctrl-C)
- Re-run FTL but without `--recreate`

Expected result:
- The lease associated with the previous attempt of the async call with expire, and the async call will be attempted again or fail, depending on the retry policy.

Actual result:
- The lease gets reaped
- The async call row loses its reference to the lease, but remains in an `executing` state forever
- This blocks the associated subscription / FSM forever

I think this happens for any case where async call leases expire, like:
- controller losing connection to the db momentarily
- controller dying

It think it **might** also happen in this case:
- controller executes `executeAsyncCalls()`
- it successfully acquires a call with `AcquireAsyncCall()
- it defers release of the call lease
- `CompleteAsyncCall` fails for any reason (eg db connection error)
Similar to what's outlined above, we now have a released lease and a async call row that will sit in an `executing` state forever

### PR DESCRIPTION
**Pull request by matt2e**
fixes #2214

Changes:
- The controller now calls the verb within the lease context, so a failure to heartbeat the lease cancels the grpc call
- A new periodic task that cleans up async calls with a state of executing but without a lease
    - This happens after the lease has been reaped in a separate job
    - Cleaning up involves:
        - Scheduling retries/catches according to the policy
        - Triggering the origin-specific code (eg: fsm async calls need the next event table to be cleared)

Known issues:
- If there is a repeating error cleaning up an async call, then it will just repeatedly fail
- Java does not currently have a way to stop execution of the call that gets canceled, so it will always continue even when the async call's lease expires

### CODE REVIEW

**Comment by alecthomas** [comment]
> * Java does not currently have a way to stop execution of the call that gets canceled, so it will always continue even when the async call's lease expires

Can you file a ticket for this?

File: backend/controller/controller.go [hunk 1]
```diff
@@ -1566,6 +1570,36 @@ func (s *Service) catchAsyncCall(ctx context.Context, logger *log.Logger, call *
...
+	defer tx.CommitOrRollback(ctx, &err)
+
+	limit := 20
+	calls, err := tx.GetZombieAsyncCalls(ctx, 20)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get zombie async calls: %w", err)
+	}
+	for _, call := range calls {
+		callResult := either.RightOf[[]byte]("async call lease expired")
+		_, err := s.dal.CompleteAsyncCall(ctx, call, callResult, func(tx *dal.Tx, isFinalResult bool) error {
```
Comment (alecthomas) [inline backend/controller/controller.go:1588]: This should be using the tx, not the dal

File: backend/controller/controller.go [hunk 2]
```diff
@@ -1566,6 +1570,36 @@ func (s *Service) catchAsyncCall(ctx context.Context, logger *log.Logger, call *
...
+			return s.finaliseAsyncCall(ctx, tx, call, callResult, isFinalResult)
+		})
+		if err != nil {
+			return 0, fmt.Errorf("failed to complete zombie async call: %w", err)
+		}
+		// TODO: telemetery
```
Comment (alecthomas) [inline backend/controller/controller.go:1594]: Let's do the telemetry in this PR, or file a ticket.

File: backend/controller/sql/queries.sql [hunk 3]
```diff
@@ -596,6 +596,27 @@ SELECT *
 FROM async_calls
 WHERE id = @id;
 
+-- name: GetZombieAsyncCalls :many
+SELECT 
+  id,
```
Comment (alecthomas) [inline backend/controller/sql/queries.sql:601]: Just use `*` here, sqlc will expand it to the individual columns.

**Review by alecthomas — APPROVED** [review]

File: backend/controller/controller.go (hunk 2 above)
Comment (matt2e) [reply backend/controller/controller.go]: Added 👍

**Pull request merged**